Persistence/DevGraph.hs

```diff
@@ -0,0 +1,1216 @@
+          (doSave, documentKey, documentLocIdBaseValue) <- case documentM of
+            Just (Entity documentKey documentLocIdBaseValue) ->
+              return (False, documentKey, documentLocIdBaseValue)
+            Nothing -> do
+              let documentLocIdBaseValue = LocIdBase
+                    { locIdBaseFileVersionId = fileVersionKey
+                    , locIdBaseKind = kind
+                    , locIdBaseLocId = locId
+                    }
+              documentKey <- insert documentLocIdBaseValue
+              insert Document
+                { documentLocIdBaseId = documentKey
+                , documentDisplayName = displayName
+                , documentName = name
+                , documentLocation = location
+                , documentVersion = version
+                }
+              return (True, documentKey, documentLocIdBaseValue)
+          let dbCache1 = addDocumentToCache libName documentKey dbCache0
+          createAllOmsOfDocument opts libEnv fileVersionKey dbCache1 doSave dGraph
+            globalAnnotations libName (Entity documentKey documentLocIdBaseValue)
+
+locIdOfDocument :: HetcatsOpts -> Maybe String -> String -> String
+locIdOfDocument opts location displayName =
+  let base = fromMaybe displayName location
+  in  if null $ libdirs opts
+      then base
+      else fromMaybe base $ stripPrefix (firstLibdir opts) base
+
+firstLibdir :: HetcatsOpts -> String
+firstLibdir opts =
+  let libdir' = head $ libdirs opts
+  in  if last libdir' == '/' then libdir' else libdir' ++ "/"
+
+-- A FileVersion only exists for the commit that actually changed the file, but
+-- the file itself also exists in future commits that changed other files.
+-- This function finds the FileVersion of the file at `location` (i.e. the
+-- commit that last changed that file).
+findFileVersion :: MonadIO m
+                => HetcatsOpts -> FileVersionId -> Maybe String
+                -> DBMonad m FileVersionId
+findFileVersion opts fileVersionKey location =
+  let isInLibdir = not (null $ libdirs opts)
+                     && isJust location
+                     && head (libdirs opts) `isPrefixOf` fromJust location
+      path = fromJust $ stripPrefix (firstLibdir opts) $ fromJust location
+      queryString = Text.pack (
+        "SELECT file_versions.id "
+        ++ "FROM file_versions "
+        ++ "INNER JOIN file_version_parents "
+        ++ "  ON file_version_parents.last_changed_file_version_id = file_versions.id "
+        ++ "INNER JOIN file_versions AS backreference "
+        ++ "  ON file_version_parents.queried_sha = backreference.commit_sha "
+        ++ "WHERE (file_versions.path = ? "
+        ++ "       AND backreference.id = ?);")
+
+  in  if isInLibdir
+      then do
+            results <-
+              rawSql queryString [ toPersistValue path
+                                 , toPersistValue fileVersionKey
+                                 ]
+            return $ head results
```

Comment:
With tracing, I could locate the following error to be caused by `results` being the empty list:
```
till@mark:~/projects/ontohub/ontohub-backend$ hets --verbose=2 --output-types=db --database-config=config/database.yml --database-subconfig=development --hets-libdirs=$HETS_LIB --database-fileversion-id=5 $HETS_LIB/Basic/RelationsAndOrders.casl
### Warning: set HETS_MAGIC to a proper magic file
Reading file /home/till/CASL/Hets-lib/Basic/RelationsAndOrders.casl
Analyzing library Basic/RelationsAndOrders version 1.0
Downloading Basic/Numbers ...
### Warning: set HETS_MAGIC to a proper magic file
Reading file /home/till/CASL/Hets-lib/Basic/Numbers.casl
Analyzing library Basic/Numbers version 1.0
Analyzing spec Basic/Numbers?Nat
Analyzing spec Basic/Numbers?Int
Analyzing spec Basic/Numbers?Rat
Analyzing spec Basic/Numbers?DecimalFraction
... loaded Basic/Numbers
Analyzing spec Basic/RelationsAndOrders?Relation
Analyzing spec Basic/RelationsAndOrders?ReflexiveRelation
Analyzing spec Basic/RelationsAndOrders?IrreflexiveRelation
Analyzing spec Basic/RelationsAndOrders?SymmetricRelation
Analyzing spec Basic/RelationsAndOrders?AsymmetricRelation
Analyzing spec Basic/RelationsAndOrders?AntisymmetricRelation
Analyzing spec Basic/RelationsAndOrders?TransitiveRelation
Analyzing spec Basic/RelationsAndOrders?SimilarityRelation
Analyzing spec Basic/RelationsAndOrders?PartialEquivalenceRelation
Analyzing spec Basic/RelationsAndOrders?EquivalenceRelation
Analyzing spec Basic/RelationsAndOrders?PreOrder
Analyzing spec Basic/RelationsAndOrders?StrictOrder
Analyzing spec Basic/RelationsAndOrders?PartialOrder
Analyzing spec Basic/RelationsAndOrders?TotalOrder
Analyzing spec Basic/RelationsAndOrders?StrictTotalOrder
Analyzing spec Basic/RelationsAndOrders?RightUniqueRelation
Analyzing spec Basic/RelationsAndOrders?LeftTotalRelation
Analyzing spec Basic/RelationsAndOrders?BooleanAlgebra
Analyzing spec Basic/RelationsAndOrders?ExtPartialOrder
Analyzing spec Basic/RelationsAndOrders?ExtTotalOrder
Analyzing spec Basic/RelationsAndOrders?ExtBooleanAlgebra
Analyzing spec Basic/RelationsAndOrders?RichPartialOrder
Analyzing spec Basic/RelationsAndOrders?RichTotalOrder
Analyzing spec Basic/RelationsAndOrders?RichBooleanAlgebra
Analyzing view Basic/RelationsAndOrders?TotalOrder_in_Nat
Analyzing view Basic/RelationsAndOrders?TotalOrder_in_Int
Analyzing view Basic/RelationsAndOrders?TotalOrder_in_Rat
Analyzing view Basic/RelationsAndOrders?PartialOrder_in_ExtBooleanAlgebra
hets: Prelude.head: empty list
```